polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -0,0 +1,1047 @@
+					collator: receipt.descriptor().collator.clone(),
+					collator_signature: receipt.descriptor().signature.clone(),
+					persisted_validation_data: persisted_validation_data.as_ref().clone(),
+					pov_hash: receipt.descriptor().pov_hash,
+					validation_code_hash: receipt.descriptor().validation_code_hash,
+				};
+
+				if Fragment::new(
+					candidate_relay_parent.clone(),
+					child_constraints,
+					prospective_candidate,
+				)
+				.is_err()
+				{
+					gum::debug!(
+						target: LOG_TARGET,
+						"Fragment::new() returned error",
+					);
+					return false
+				}
+			}
+
+			// If we got this far, it can be added to the chain right now.
+			true
+		} else {
+			// Otherwise it is or can be an unconnected candidate.
+			true
+		}
+	}
+
+	/// Select `count` candidates after the given `ancestors` which pass
+	/// the predicate and have not already been backed on chain.
+	///
+	/// The intention of the `ancestors` is to allow queries on the basis of
+	/// one or more candidates which were previously pending availability becoming
+	/// available or candidates timing out.
+	pub(crate) fn find_backable_chain(
+		&self,
+		ancestors: Ancestors,
+		count: u32,
+		pred: impl Fn(&CandidateHash) -> bool,
+	) -> Vec<CandidateHash> {
+		if count == 0 {
+			return vec![]
+		}
+		let base_pos = self.find_ancestor_path(ancestors);
+
+		let actual_end_index = std::cmp::min(base_pos + (count as usize), self.chain.len());
+		let mut res = Vec::with_capacity(actual_end_index - base_pos);
+
+		for elem in &self.chain[base_pos..actual_end_index] {
+			if self.scope.get_pending_availability(&elem.candidate_hash).is_none() &&
+				pred(&elem.candidate_hash)
+			{
+				res.push(elem.candidate_hash);
+			} else {
+				break
+			}
+		}
+
+		res
+	}
+
+	// Tries to orders the ancestors into a viable path from root to the last one.
+	// Stops when the ancestors are all used or when a node in the chain is not present in the
+	// ancestor set. Returns the index in the chain were the search stopped.
+	fn find_ancestor_path(&self, mut ancestors: Ancestors) -> usize {
+		if self.chain.is_empty() {
+			return 0;
+		}
+
+		for (index, candidate) in self.chain.iter().enumerate() {
+			if !ancestors.remove(&candidate.candidate_hash) {
+				return index
+			}
+		}
+
+		// This means that we found the entire chain in the ancestor set. There won't be anything
+		// left to back.
+		self.chain.len()
+	}
+
+	// Return the earliest relay parent a new candidate can have in order to be added to the chain.
+	// This is the relay parent of the last candidate in the chain.
+	// The value returned may not be valid if we want to add a candidate pending availability, which
+	// may have a relay parent which is out of scope. Special handling is needed in that case.
+	fn earliest_relay_parent(&self) -> RelayChainBlockInfo {
+		if let Some(last_candidate) = self.chain.last() {
+			self.scope
+				.ancestor_by_hash(&last_candidate.relay_parent())
```

Comment:
Why the relay parent of the last candidate is not good in general ?

Reply:
if the candidate is pending availability, we keep its relay parent in `self.pending_availability`.
We need to differentiate between pending candidates and new candidates because pending candidates can have a relay parent which is out of scope (older than allowed_ancestry_len)